### PR TITLE
Add malformed feature configuration event

### PR DESCRIPTION
**Pull request by jhugman**
Fixes [EXP-3310](https://mozilla-hub.atlassian.net/browse/EXP-3310).

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

**Comment by jhugman** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) **What questions will you answer with this data?**
- What parts of configurations from experiments and rollouts are semantically (according to the app's code) invalid, or malformed.

2) **Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?**
- This will generalizes recording of events for breakages caused in all features as described by Nimbus. It also provides us opportunities to build out leading indicators of misconfigured or broken features due to rollouts or experiments (e.g. dashboards in Experimenter).

3) **What alternative methods did you consider to answer these questions? Why were they not sufficient?**
- Currently, each feature has needed to record its own events for broken features. This is insufficient because it requires custom analysis, and additional event types.

4) **Can current instrumentation answer these questions?**
- Yes, but this new event replaces this previous instrumentation.

5) **List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data collection ](https://wiki.mozilla.org/Firefox/Data_Collection)[categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.**
- All data is Category 1.

6) **How long will this data be collected?**
- This will not expire.

7) **What populations will you measure?**
-  All release, beta, and nightly users with telemetry enabled.

8) **Please provide a general description of how you will analyze this data.**
- This will enable tooling to be built into Experimenter, using Glean.

9) **Where do you intend to share the results of your analysis?**
-  Only on Glean, Amplitude and with mobile product teams.

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt [hunk 1]
```diff
@@ -437,26 +441,45 @@ open class Nimbus(
...
-    @WorkerThread
+    @AnyThread
     internal fun recordExposureOnThisThread(featureId: String) = withCatchAll("recordExposure") {
-        val activeExperiments = getActiveExperiments()
-        activeExperiments.find { it.featureIds.contains(featureId) }?.also { experiment ->
-            NimbusEvents.exposure.record(NimbusEvents.ExposureExtra(
-                experiment = experiment.slug,
-                branch = experiment.branchSlug,
+        // First, we get the enrolled feature, if there is one, for this id.
+        val enrollment = nimbusClient.getEnrollmentByFeature(featureId) ?: return@withCatchAll
```
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:450]: Rewritten `recordExposureEvent` in terms of `getEnrollmentByFeature`. Tests still pass.

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/NimbusInterface.kt [hunk 2]
```diff
@@ -157,34 +157,6 @@ interface NimbusInterface : FeaturesInterface, GleanPlumbInterface, NimbusEventS
...
-     * are fetched.
-     *
-     * This function is safe to call even when there is no active experiment for the feature. The SDK
-     * will ensure that an event is only recorded for active experiments.
-     *
-     * @param featureId string representing the id of the feature for which to record the exposure
-     *     event.
-     */
-    override fun recordExposureEvent(featureId: String) = Unit
-
```
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/NimbusInterface.kt:187]: Moved to `FeaturesInterface`.

File: components/nimbus/ios/Nimbus/Nimbus.swift [hunk 3]
```diff
@@ -99,23 +99,41 @@ extension Nimbus: NimbusEventStore {
...
-        // Next, we search for any experiment that has a matching featureId. This depends on the
-        // fact that we can only be enrolled in a single experiment per feature, so there should
-        // only ever be zero or one experiments for a given featureId.
-        if let experiment = activeExperiments.first(where: { $0.featureIds.contains(featureId) }) {
+        // First, we get the enrolled feature, if there is one, for this id.
+        if let enrollment = getEnrollmentByFeature(featureId: featureId),
+           // If branch is nil, this is a rollout, and we're not interested in recording
+           // exposure for rollouts.
+           let branch = enrollment.branch
+        {
```
Comment (jhugman) [inline components/nimbus/ios/Nimbus/Nimbus.swift:107]: Rewritten `recordExposureEvent` in terms of simpler and cheaper `getEnrollmentByFeature`. Tests still pass.

File: components/nimbus/metrics.yaml [hunk 4]
```diff
@@ -173,6 +173,36 @@ nimbus_events:
...
+        description: The identifier of the feature that is has a malformed
+          configuration
+      part_id:
+        type: string
+        description: The identifier of the feature-specific part that is
+          malformed. e.g. the card or message id.
+    bugs:
+      - https://jira.mozilla.com/browse/EXP-3310
+    data_reviews:
+      - https://github.com/mozilla/application-services/pull/5440#issuecomment-1481999207
```
Comment (jhugman) [inline components/nimbus/metrics.yaml:199]: Pending data review.

**Review by travis79 — APPROVED** [review]
> # Request for data collection review form
> 
> **All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**
> 
>     1. **What questions will you answer with this data?**
> 
> 
>     * What parts of configurations from experiments and rollouts are semantically (according to the app's code) invalid, or malformed.
> 
> 
>     2. **Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?**
> 
> 
>     * This will generalizes recording of events for breakages caused in all features as described by Nimbus. It also provides us opportunities to build out leading indicators of misconfigured or broken features due to rollouts or experiments (e.g. dashboards in Experimenter).
> 
> 
>     3. **What alternative methods did you consider to answer these questions? Why were they not sufficient?**
> 
> 
>     * Currently, each feature has needed to record its own events for broken features. This is insufficient because it requires custom analysis, and additional event types.
> 
> 
>     4. **Can current instrumentation answer these questions?**
> 
> 
>     * Yes, but this new event replaces this previous instrumentation.
> 
> 
>     5. **List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data collection ](https://wiki.mozilla.org/Firefox/Data_Collection)[categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.**
> 
> 
>     * All data is Category 1.
> 
> 
>     6. **How long will this data be collected?**
> 
> 
>     * This will not expire.
> 
> 
>     7. **What populations will you measure?**
> 
> 
>     * All release, beta, and nightly users with telemetry enabled.
> 
> 
>     8. **Please provide a general description of how you will analyze this data.**
> 
> 
>     * This will enable tooling to be built into Experimenter, using Glean.
> 
> 
>     9. **Where do you intend to share the results of your analysis?**
> 
> 
>     * Only on Glean, Amplitude and with mobile product teams.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/).

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, through the "Send Usage Data" preference in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

Permanent collection to be monitored over time by jhugman.

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 1, Technical data

5) Is the data collection request for default-on or default-off?

Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice?

Yes

8) Does the data collection use a third-party collection tool?

No

## Result

data-review+

File: components/nimbus/metrics.yaml (hunk 4 above)
Comment (jhugman) [reply components/nimbus/metrics.yaml]: Added correct link.

**Pull request merged**